aesara/scalar/math.py

```diff
@@ -1070,3 +1073,224 @@ def c_code_cache_version(self):
 
 
 softplus = Softplus(upgrade_to_float, name="scalar_softplus")
+
+
+class BetaInc(TernaryScalarOp):
+    """
+    Regularized incomplete beta function
+    """
+
+    nfunc_spec = ("scipy.special.betainc", 3, 1)
+
+    def impl(self, a, b, x):
+        return scipy.special.betainc(a, b, x)
+
+    def grad(self, inp, grads):
+        a, b, x = inp
+        (gz,) = grads
+
+        return [
+            gz * betainc_dda_scalar(a, b, x),
+            gz * betainc_ddb_scalar(a, b, x),
+            gz
+            * exp(
+                log1p(-x) * (b - 1)
+                + log(x) * (a - 1)
+                - (gammaln(a) + gammaln(b) - gammaln(a + b))
+            ),
+        ]
+
+
+betainc = BetaInc(upgrade_to_float_no_complex, name="betainc")
+
+
+class BetaIncDda(TernaryScalarOp):
+    """
+    Gradient of the regularized incomplete beta function wrt to the first argument (a)
+    """
+
+    def impl(self, a, b, x):
+        return _betainc_derivative(a, b, x, wrtp=True)
+
+
+betainc_dda_scalar = BetaIncDda(upgrade_to_float_no_complex, name="betainc_dda")
+
+
+class BetaIncDdb(TernaryScalarOp):
+    """
+    Gradient of the regularized incomplete beta function wrt to the second argument (b)
+    """
+
+    def impl(self, a, b, x):
+        return _betainc_derivative(a, b, x, wrtp=False)
+
+
+betainc_ddb_scalar = BetaIncDdb(upgrade_to_float_no_complex, name="betainc_ddb")
+
+
+def _betainc_derivative(p, q, x, wrtp=True):
```

Comment:
Is the/a Aesara implementation of `_betainc_derivative` prohibitively slow compared to this Python implementation?  

If not, we should definitely use the Aesara implementation.  If so, we need to figure out why and open a independent line of investigation for why that's the case, and fix it.

Reply:
I actually did not write it! 

The BetaInc code which was just a copy of the c code behind the scipy.special.betainc was prohibitively slow (to compile specially), as were the autodiff derivatives obtained from it. I have some crude benchmarks here: https://github.com/ricardoV94/derivatives_betainc/blob/master/comparison_aesara.ipynb

If we want a test case for exploring the slowness of scan that seems like a good start, as we have the scipy vs aesara with the exact same algorithm under the hood.

The derivatives are a complete different algorithm so they might be fine. Do you think it's worth exploring that? I guess the concern here is that they break the auto-diff chain? Or is it an issue for the numbat backed as is?